### ISSUE TITLE
comma-dangle always-multiline should only count newlines before closing curly

### ISSUE THREAD

**Issue by steadicat** [eslint/eslint#2091]
The new `always-multiline` behavior of `comma-dangle` requires a trailing comma if a newline appears anywhere in the literal. It would probably make more sense if it only considered the space right before the closing bracket.

E.g. these two cases are currently not allowed, but they probably should be:

```javascript
{x: {
  foo: 'bar',
}}
```

```javascript
{a: 1, b: 2,
  c: 3, d: 4}
```

This is a followup to #1984.

**Comment by nzakas** [comment]
Can you paste the output you get when running each of those examples?

In general, when filing issues it's useful to include the source code, your configuration, and the output. Really helps us triage.

**Comment by steadicat** [comment]
```javascript
// test.js
var a = {x: {
  foo: 'bar',
}};
var b = {a: 1, b: 2,
         c: 3, d: 4};
```
```javascript
// config.json
{
    "rules": {
        "comma-dangle": [2, "always-multiline"]
    }
}
```
```
$ eslint --reset --no-eslintrc --config config.json test.js

test.js
  3:1   error  Missing trailing comma  comma-dangle
  5:19  error  Missing trailing comma  comma-dangle

✖ 2 problems (2 errors, 0 warnings)
```

**Comment by nzakas** [comment]
So you're saying that the rule should ignore these situations because...?

**Comment by steadicat** [comment]
Because the following looks really weird and I doubt this was the intent of `always-multiline`:

```javascript
var a = {x: {
  foo: 'bar',
},};
var b = {a: 1, b: 2,
         c: 3, d: 4,};
```

`always-multiline` makes sense if you want to enforce commas at the end of each line. I can't think of why anyone would want a line-break _anywhere in the literal_ to cause the last property to need a dangling comma.

**Comment by nzakas** [comment]
I'm just trying to come up with an explanation that can be turned into code. Part of me thinks that if you use weird formats like this, you are voiding your warranty on having a linter understand it. The other part of me is trying to accept that there's an exception that would make sense for these samples and wouldn't be unreasonable to implement. I'm asking for your help with that, so I need more than "this would be weird" (because another option would be to use a more normalized style to appease the rule).

**Comment by steadicat** [comment]
I think the rule that can be turned into code is pretty simple, and is what what originally proposed in #1984: require a dangling comma if and only if there is a newline between the last value and the closing parenthesis.

Here's the reasoning behind this. The goal is to enforce a comma at the end of each line:

```javascript
var x = {
  a: 1,
  b: 2,
  c: 3,
  d: 4,
};
```

Typically when you split an object or array, you place one value/property per line. The goal is to make all lines look alike, and to make it so a code modification that appends one element to only have to touch one line.

This formulation of the rule takes care of a bunch of cases where the rule as implemented today doesn't make sense. For example, I can't think of why anyone would ever want a dangling comma in this case:

```javascript
var x = [{
  a: 1,
  b: 2,
  c: 3,
  d: 4,
},];
```

Any time there is no new line before the closing parenthesis, the dangling comma doesn't make sense:

```javascript
var x = {
  a: 1,
  b: 2,
  c: 3,
  d: 4,};
```

This is a weird edge case, but I don't think anyone would argue that a comma is better than no comma in this case. The last line is already different from the others, so there is no reason for it to have a comma. And nobody's used to seeing dangling commas right before closing parenthesis on the same line.

Does this make sense?

**Comment by nzakas** [comment]
Ah, got it, thanks for explaining. I guess my hope is always that ESLint will help people recognize how strange their code style is and adapt accordingly. I suppose that's unrealistic. :)

Seems like a simple enough change.

**Comment by steadicat** [comment]
Glad I could help. I'm curious though, are you saying the `always-multiline` comma dangle is strange? I guess if you're talking about the current rule (anywhere in the literal), then I agree with you. The way I'm proposing, I find that much less strange than `always` (who likes `{x:2,}`?!?).

Btw, the multiline behavior I'm describing is part of the code standards at Facebook, so I'm not making it up. :grin:

**Comment by nzakas** [comment]
The style of the code is very strange, regardless if it comes from Facebook or anywhere else. :)

**Comment by mhart** [comment]
:+1:  on this – I don't think it's a "very" strange style – and I think you've hit the nail on the head that the `always-multiline` should be about the ability to copy/paste/shift-lines around – ie, for cases where the comma is the last character on the line.

**Comment by mathieumg** [comment]
> always-multiline should be about the ability to copy/paste/shift-lines around – ie, for cases where the comma is the last character on the line.

Exactly the same opinion over here. :+1: I guess it means we're a bunch of weirdos! :stuck_out_tongue_winking_eye:

**Comment by matthewwithanm** [comment]
Another example:

```js
new Map([
  [key, {
    a: 1,
    b: 2,
    c: 3,
  }],
]);
```

**Comment by BenoitZugmeyer** [comment]
Working on this.
